src/Microsoft.AspNet.Razor/CodeGenerators/CSharpTagHelperCodeRenderer.cs

```diff
@@ -249,70 +206,135 @@ protected virtual string GenerateUniqueId()
+            {
+                var associatedDescriptors = tagHelperDescriptors.Where(descriptor =>
+                    descriptor.Attributes.Any(attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key)));
+
+                // Bound attributes have associated descriptors.
+                if (associatedDescriptors.Any())
+                {
+                    if (!renderedBoundAttributeNames.Add(attribute.Key))
                     {
+                        // First attribute value wins if there are duplicates.
                         continue;
                     }
 
-                    // We need to inform the context of the attribute value.
-                    _writer
-                        .WriteStartInstanceMethodInvocation(
-                            ExecutionContextVariableName,
-                            _tagHelperContext.ExecutionContextAddTagHelperAttributeMethodName)
-                        .WriteStringLiteral(attributeName)
-                        .WriteParameterSeparator()
-                        .Write(valueAccessor)
-                        .WriteEndMethodInvocation();
+                    // We need to capture the tag helper's property value accessor so we can retrieve it later
+                    // if there are more tag helpers that need the value.
+                    string valueAccessor = null;
+
+                    foreach (var associatedDescriptor in associatedDescriptors)
+                    {
+                        var associatedAttributeDescriptor = associatedDescriptor.Attributes.First(
+                            attributeDescriptor => attributeDescriptor.IsNameMatch(attribute.Key));
+                        var tagHelperVariableName = GetVariableName(associatedDescriptor);
+
+                        valueAccessor = RenderBoundAttribute(
+                            attribute.Key,
+                            attribute.Value,
+                            tagHelperVariableName,
+                            valueAccessor,
+                            associatedAttributeDescriptor);
+                    }
                 }
                 else
                 {
-                    // The attribute value has already been recorded, lets retrieve it from the stored value
-                    // accessors.
-                    _writer
-                        .WriteStartAssignment(valueAccessor)
-                        .Write(htmlAttributeValues[attributeName])
-                        .WriteLine(";");
+                    RenderUnboundAttribute(attribute.Key, attribute.Value);
+                }
+            }
+        }
+
+        private string RenderBoundAttribute(
+            string attributeName,
+            Chunk attributeValueChunk,
+            string tagHelperVariableName,
+            string previousValueAccessor,
+            TagHelperAttributeDescriptor attributeDescriptor)
+        {
+            if (attributeValueChunk == null)
```

Comment:
This check should be outside the `foreach` loop at line 248. No need to do something general for every associated tag helper.